### ISSUE TITLE
Compilation Failing

### ISSUE THREAD

**Issue by rlrangel** [KratosMultiphysics/Kratos#11062]
I have just updated Master branch and I am now getting the following error when compiling it (also shown in image below):

_C:\Users\rrangel\Desktop\Repos\Kratos\kratos\processes\apply_ray_casting_interface_recognition_process.cpp(74,73): error C2946: explicit instantiation; 'Kratos::ApplyRayCastingInterfaceRecognitionProcess<3>::<lambda_ce7114c54fe906ddd2f12c6de41d62e9>' is not a template-class specialization [C:\Users\rrangel\Desktop\Repos\Kratos\build\Debug\kratos\KratosCore.vcxproj]_

_C:\Users\rrangel\Desktop\Repos\Kratos\kratos\processes\apply_ray_casting_process.cpp(565,53): error C2946: explicit instantiation; 'Kratos::ApplyRayCastingProcess<2>::<lambda_5d12caca6125160d56f841004477be35>' is not a template-class specialization [C:\Users\rrangel\Desktop\Repos\Kratos\build\Debug\kratos\KratosCore.vcxproj]_

It happens both in Debug and Release modes with any application that I try to compile.

If I go back to the previous state of two weeks ago, it compiles normally.

I tried to delete Kratos and clone it again, but the error persists.

I'm using Windows 10, Python 3.9, Cmake 3.18, Boost 1.74.

![image](https://user-images.githubusercontent.com/69159206/234843151-c44aa773-bfbe-4f1c-bfcf-1f0c8103322a.png)

**Comment by ddiezrod** [comment]
Hmm its weird that it compiles fine in the rest of the machines , which visual studio version are you using?

**Comment by ddiezrod** [comment]
Even weirder as it the same compiler we use in our CI... I honestly don't know how to debug this without being able to reproduce it. Do you have any clue? @matekelemen @roigcarlo

**Comment by roigcarlo** [comment]
I will take a look

**Comment by roigcarlo** [comment]
Mm I am afraid I can't also reproduce it, but I might have a slight idea about what is happening.

The error seems to be caused by a bug in the compiler. For more info read:
https://developercommunity.visualstudio.com/t/msvc-instantation-issue/1059795
https://jeffpar.github.io/kbarchive/kb/263/Q263633/

Could you try a couple of changes:

At the bottom of the file try to change 

```C++

    template class Kratos::ApplyRayCastingProcess<2>;
    template class Kratos::ApplyRayCastingProcess<3>;

}  // namespace Kratos.
```

By either:
1 - (Remove `Kratos::`)

```C++

    template class ApplyRayCastingProcess<2>;
    template class ApplyRayCastingProcess<3>;

}  // namespace Kratos.
```

2 - Move the instantiation outside the namespace
```C++

}  // namespace Kratos.

template class Kratos::ApplyRayCastingProcess<2>;
template class Kratos::ApplyRayCastingProcess<3>;
```

If nothing of this work I am afraid you will need to update to a more recent version of VS.

For @KratosMultiphysics/technical-committee, seems that we were unawarely using a very poorly tested c++17 feature when implemented the registry (we already found problems in clang 12< and MSVC). If errors like this arise tell me and I will take care of upgrading the versions of the compilers accordingly.

**Comment by rlrangel** [comment]
Thank you @roigcarlo !
I will try these suggestions soon, and then I tell if it worked.

**Comment by rlrangel** [comment]
Those changes didn't work.
I will update VS to a more recent version then.

**Comment by roigcarlo** [comment]
Perfect!

**Comment by rlrangel** [comment]
I am reopening the issue as I am facing a similar problem again after updating the Master branch.

I am using the same Windows 10 machine as before, with Python 3.9, Boost 1.74, CMake 3.27, and Visual Studio 2019 version 16.11.

Some of the errors that appear in the command window are shown below.

Any ideas?



_C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\type_traits(1613,97): error C2968: '_Is_invocable_r<std::shared_ptr<Kratos::ApplyRayCastingProcess<2> > >': recursive alias declaration (com
piling source file C:\Users\rrangel\Desktop\Repos\Kratos\kratos\processes\apply_ray_casting_process.cpp) [C:\Users\rrangel\Desktop\Repos\Kratos\build\Debug\kratos\KratosCore.vcxproj]

C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\type_traits(1613,97): error C2968: '_Is_invocable_r<std::shared_ptr<Kratos::ApplyRayCastingProcess<2> > >': recursive alias declaration (com
piling source file C:\Users\rrangel\Desktop\Repos\Kratos\kratos\processes\apply_ray_casting_interface_recognition_process.cpp) [C:\Users\rrangel\Desktop\Repos\Kratos\build\Debug\kratos\KratosCore.vcxproj]

C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\type_traits(1692,69): error C2938: 'std::_Is_invocable_r_' : Failed to specialize alias template (compiling source file C:\Users\rrangel\Des
ktop\Repos\Kratos\kratos\processes\apply_ray_casting_process.cpp) [C:\Users\rrangel\Desktop\Repos\Kratos\build\Debug\kratos\KratosCore.vcxproj]

C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\type_traits(1692,69): error C2938: 'std::_Is_invocable_r_' : Failed to specialize alias template (compiling source file C:\Users\rrangel\Des
ktop\Repos\Kratos\kratos\processes\apply_ray_casting_interface_recognition_process.cpp) [C:\Users\rrangel\Desktop\Repos\Kratos\build\Debug\kratos\KratosCore.vcxproj]

C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\type_traits(1692,69): error C3770: 'unknown-type': is not a valid base class (compiling source file C:\Users\rrangel\Desktop\Repos\Kratos\kr
atos\processes\apply_ray_casting_interface_recognition_process.cpp) [C:\Users\rrangel\Desktop\Repos\Kratos\build\Debug\kratos\KratosCore.vcxproj]

C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\type_traits(1692,69): error C3770: 'unknown-type': is not a valid base class (compiling source file C:\Users\rrangel\Desktop\Repos\Kratos\kr
atos\processes\apply_ray_casting_process.cpp) [C:\Users\rrangel\Desktop\Repos\Kratos\build\Debug\kratos\KratosCore.vcxproj]_

**Comment by matekelemen** [comment]
I'm guessing the problem is that explicit instantiations of the class are registered within the declaration of the class
```cpp
template<std::size_t TDim = 3>
class KRATOS_API(KRATOS_CORE) ApplyRayCastingProcess : public Process
{
    /*...*/
    KRATOS_REGISTRY_ADD_TEMPLATE_PROTOTYPE("Processes.KratosMultiphysics", ApplyRayCastingProcess, 2)
    KRATOS_REGISTRY_ADD_TEMPLATE_PROTOTYPE("Processes.KratosMultiphysics", ApplyRayCastingProcess, 3)
    KRATOS_REGISTRY_ADD_TEMPLATE_PROTOTYPE("Processes.All", ApplyRayCastingProcess, 2)
    KRATOS_REGISTRY_ADD_TEMPLATE_PROTOTYPE("Processes.All", ApplyRayCastingProcess, 3)
    /*...*/
};
```
which creates an infinite loop, always trying to instantiate `TDim=3` from `TDim=2` and vise versa. @roigcarlo can `KRATOS_REGISTRY_ADD_TEMPLATE_PROTOTYPE` go outside the class definition? If not, a specific class should always register itself only, and not other instantiations.
```diff
    template<std::size_t TDim = 3>
class KRATOS_API(KRATOS_CORE) ApplyRayCastingProcess : public Process
{
    /*...*/
-    KRATOS_REGISTRY_ADD_TEMPLATE_PROTOTYPE("Processes.KratosMultiphysics", ApplyRayCastingProcess, 2)
-    KRATOS_REGISTRY_ADD_TEMPLATE_PROTOTYPE("Processes.KratosMultiphysics", ApplyRayCastingProcess, 3)
-    KRATOS_REGISTRY_ADD_TEMPLATE_PROTOTYPE("Processes.All", ApplyRayCastingProcess, 2)
-    KRATOS_REGISTRY_ADD_TEMPLATE_PROTOTYPE("Processes.All", ApplyRayCastingProcess, 3)
+    KRATOS_REGISTRY_ADD_TEMPLATE_PROTOTYPE("Processes.KratosMultiphysics", ApplyRayCastingProcess, TDim)
+    KRATOS_REGISTRY_ADD_TEMPLATE_PROTOTYPE("Processes.All", ApplyRayCastingProcess, TDim)
    /*...*/
};
```

**Comment by roigcarlo** [comment]
Mmm no we not do that, we need the exact instance so gets added to the registry (otherwise TDim appers in the registry as the key for both instances)

Let me take a look, it was not crashing in the CI or in my machines. It could be something related with an app not being compiled which is including the file in a way I did not consider.

@rlrangel could you send me the list of applications that you have enabled and the version of Visual studio that you are using?

**Comment by AlejandroCornejo** [comment]
In one of my pc with Windows i was getting the same compilation errors reported by @rlrangel

**Comment by AlejandroCornejo** [comment]
I was compiling in my case the LinearSolvers, StructuralMechApp and the ConstitutiveLawsApp. VS 2019 c++17

**Comment by rlrangel** [comment]
> @rlrangel could you send me the list of applications that you have enabled and the version of Visual studio that you are using?

Visual Studio Community 2019 Version 16.11.26

I am trying to compile the DelaunayMeshingApplication and ThermalDEMApplication applications.
bat file below:

```
rem Set compiler
set CC=cl.exe
set CXX=cl.exe

rem Set variables
set KRATOS_SOURCE=~0,-1%/..
set KRATOS_BUILD=%KRATOS_SOURCE%/build
set KRATOS_APP_DIR=applications

rem Set basic configuration
set KRATOS_BUILD_TYPE=Debug
set BOOST_ROOT=C:\Program Files\Boost\Boost_174
set PYTHON_EXECUTABLE=C:\Program Files\Python\Python_39\python.exe

rem Set applications to compile
set KRATOS_APPLICATIONS=
CALL :add_app %KRATOS_APP_DIR%\DelaunayMeshingApplication;
CALL :add_app %KRATOS_APP_DIR%\ThermalDEMApplication;

rem Clean
del /F /Q "%KRATOS_BUILD%\%KRATOS_BUILD_TYPE%\cmake_install.cmake"
del /F /Q "%KRATOS_BUILD%\%KRATOS_BUILD_TYPE%\CMakeCache.txt"
del /F /Q "%KRATOS_BUILD%\%KRATOS_BUILD_TYPE%\CMakeFiles"

rem Configure
@echo on
cmake -G"Visual Studio 16 2019" -H"%KRATOS_SOURCE%" -B"%KRATOS_BUILD%\%KRATOS_BUILD_TYPE%"  ^
-DUSE_EIGEN_MKL=OFF                                       ^
-DUSE_TRIANGLE_NONFREE_TPL=ON                             ^
-DUSE_TETGEN_NONFREE_TPL=ON                               ^
-DCMAKE_CXX_FLAGS=" -D_SCL_SECURE_NO_WARNINGS /MP "

rem Build
cmake --build "%KRATOS_BUILD%/%KRATOS_BUILD_TYPE%" --target install -- /property:configuration=%KRATOS_BUILD_TYPE% /p:Platform=x64
goto:eof

rem Function to add apps
:add_app
set KRATOS_APPLICATIONS=%KRATOS_APPLICATIONS%%1;
goto:eof
```

**Comment by roigcarlo** [comment]
I am still unable to reproduce this.

To elaborate more on the recursive template issue, that should never happen. The template instantiations happens inside a lambda function the result of which is assigned to a static member variable, the compiler should evaluate the content of the lambda after the class definition is generated.

As I see it the code should generate as follows:
- Definition of the class parsed by the compiler
- `ApplyRayCastingProcess.cpp` is parsed
- `ApplyRayCastingProcess<2>` explicit instantiation is found
- Code is generated for `ApplyRayCastingProcess<2>`
- Instantiation of `ApplyRayCastingProcess<3>` does NOT happen because it does not need the definition to generate the code of the lambda (only the type name)
- Code is generated for `ApplyRayCastingProcess<3>` (Same applies here)
- During linking both definitions of `ApplyRayCastingProcess<2>` and `ApplyRayCastingProcess<3>` are available. This will cause that
- Once the program is loaded, the lambda gets evaluated as the static inline variable needs to have a value -> The registration happens.

Super simple code that basically does the same ( I included void calls to mimic the effect of the registry.register but they don't do anything here)
https://godbolt.org/z/bh5dfn7vr

If it were to instantiate the other instance (2) while generating the code for (3), it firstly should complain avoid the use of a partial class (which is indeed invalid). If this is what is happening then I can only assume that there is a problem in that specific version of the MSVC compiler, as this code is indeed a corner case of some recent implementations.

A possible solution of this if @KratosMultiphysics/technical-committee agrees is to move the registry macro of template classes oustise its declaration (aka, having to macros).

**Comment by matekelemen** [comment]
> - Instantiation of ApplyRayCastingProcess<3> does NOT happen because it does not need the definition to generate the code of the lambda (only the type name)

Is that guaranteed by the standard, or is it just how compilers behave right now? Honestly, I'm a bit out of my depth here because I have no idea what the standard evaluation order is supposed to be with `inline` members. Hopefully, the standard does guarantee what you're saying and it's just an MSVC bug again.

Btw, your example does indeed [fail with MSVC <= v19.34](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxCAAzBqkAA6oCoRODB7evnrJqY4CQSHhLFEx8baY9vkMQgRMxASZPn5cFVXptfUEhWGR0XEJCnUNTdmtQ109xaUDAJS2qF7EyOwc5rHByN5YANQmsW5UXgwO6WL72CYaAILrm9uYewdOQ8SYrBdXt2YbJw9PbgAbpgHCRPjc7n8vLt9m42CwSABPcG3G5DdAgEDA0HEWHBAgXHZUJgAa0wAH03sA8K9EU8ACJ7ADsViZ9P2VghNwImBYiQMPIB%2BJ2ABUUa8vA4dkIviYWV8doqdhNHMgdsFaMFHsK3govLQCAy9gBWKzG%2BkQWY7AC0hOFcs51yVzqJpIpVJpBGIiIAdIkvAoEOSIkxkCSICLZhyFS7FejMWgvIbYbC9mYzNhAWIvExHEYdgQEI80LtUFQC0XlXVVTt4aUiSRRft2emAamRW2DlWMSBXLRozdY4qA8FgKKjidqk7m93MRPTgJYfGQIH6ph0OTEl7YTKDmZPrFsJaD46hzsR/mRfPqhYjcvr%2BklwQe6u3hut7iDru3LED0eo4eHJ7IOQ5XscC4ME66BMEaJimnBFpRiybwEEsDCziALBuuSr7rjusL7oBh7HrErLsqRMaxmBk7pLe6ARLB8HmseyGYKhxDocuWFkjhCBrug%2BEHL%2BRH/tGbIDk6Q4oWhTakcBkmKnKiESUpKncry/K5o8EpSt%2BhGXBR6l8gK2lepKhrfsJBmOl8wpYcElrMqeSr8MQOwQEwSaoOqPIsCAro8R6tJWg6lEusuibJgcqaELynZuBhfYSc6qkQuJNwcPMtCcMavB%2BBwWikKgnBuNY1jKosyyPOsPCkAQmiZfMJIgJIAAcPoaK1sTpgAnK1xpcF1XAAGxMsa%2BicJIvB%2BVwGgJPlhXFRwvAKCACT1QVmWkHAsAwIgvaqCCXlkBQHnEMACjKIYlRCAgqAAO75bVaB8nQubpFdIS0LdD0LbwL2JHQ/TAFwsStADQPEKErCrP9qCvfQxAAPJJj9j0NaQmCHcg1znaty2Y9jtT4PlvD8IIIhiOwUgyIIigqOom2kLorQGEYKBlZY%2Bh4BEq2QPMqBbuk%2BO8KgwLEMQeBYHzloLEsKx6OiwSfTdd3o9wvD3cQTCJJwPBZTleUY0t2DY8dOyqK1w3WsNkg7MAyBqqDPpcO5pWWNYpA7LghCNjVsy8BtWizE1LVMj6xrGj1PVmJ1PxcEysTDbEE0cFNpB%2BWY41/UVnArWtdUNfMO37RDiPkJQZf9ICyCJIk5KAlwPXkrEkjkpbw2kFggJ4CsABqeCYPdSOJIwetk3QPLEKtEARBjETBPUiLj6QC/MN6SMRNoIIbc98NsIISMMLQy9M1gEReMAbhiLQ%2BO1VgWFGOIZ94G8pzAvjhVY0dPIr/ilQY01BEbW3oPBYAxl6PAflNrzCoAYC6A8h4jzHhrWmwhRDiBpuTeQSg1AYxZvoQwxhOY2CATLAWQsBD4wAPTxlMB7SwCdRbi0ltLeA8w7A73SC4Bg7hPDNBauNQIWppj9BTrkNIAhRh%2BEkONCR1Qph9DKG0LhAhOgjH4dkWRKiILqO6CIpRcRbDDEaJomR40JgNEUSUMRHDKoKy4AbDguVSA5yWhbK2tYFA1x2I3HqPpW7uR9kQNy/tA5F3mEWJgWAYiy2ymnaaIAs6uONnnWwBcg5bRLhAJAiwCD%2BgIBXCA65fasL0NgymmDpDYPpngpmBCtY63Hk4lxbjOAo3yUmHYZYPHDS8T4vxAS7YQA8AjaIaYwYB0LjAyJ7wYmUCcenPyxp5qpIJqtdaETSDNQGj6HqI0mRMg0FIVqI0urJPibEI2TMlrhJmanMw1zFppMySHUg4tUjOEkEAA%3D).

**Comment by roigcarlo** [comment]
> Is that guaranteed by the standard, or is it just how compilers behave right now?
I will look on to it, I was under the impression it is guaranteed.

**Comment by rubenzorrilla** [comment]
Closing as it should be fixed in #11345.

**Comment by maceligueta** [comment]
Reopening as I am getting the same error as posted in the first message of this issue... 
Visual Studio 19 Community, too.
It seems the pull requests only fixed part of the problem.

**Comment by maceligueta** [comment]
Apparently this is related to some kind of bug in the MS compiler. Closing, as it is a known problem.

**Comment by matekelemen** [comment]
@maceligueta can you please link the bug report or discussion where you found out about it?

**Comment by maceligueta** [comment]
Well, this conclusion came up in a private conversation with @roigcarlo , in which he mentioned that this problem disappears when the VS compiler is updated, so I thought it had to be a compiler problem. I did not look for a confirmation of this...

**Comment by roigcarlo** [comment]
@matekelemen He was having the C2946 error (the former one, which is detailed here https://developercommunity.visualstudio.com/t/msvc-instantation-issue/1059795) nothing to do with the alias.

As a recap, for the future. This issue tackles two problems:
- **C2946**: Problem while trying to import a template specialization containing a lambda (which was fixed https://developercommunity.visualstudio.com/t/msvc-instantation-issue/1059795, in 16.8)
- **C2968**: Problem because we were trying to instantiate the specialization of DIM=2 and DIM=3 in code generation of both DIM=2 and DIM=3 classes, which caused a deadlock / missing definition if the compiler tried to generate the code of the template instance while in the lambda. As I understand the lambda code should be generated AFTER the class code (what gcc and clang must be doing becuase otherwise would have also failed), would have prevented this, but I could not find any documentation supporting this claim, so we changed how that worked in #11345 so only the specialized instance for which the code is being generated appears inside the lambda definition
